### PR TITLE
dropSamples plays well with optimizer

### PR DESCRIPTION
**Pull request by danking**
This should play well with [the optimizer](https://github.com/hail-is/hail/blob/master/src/main/scala/is/hail/expr/Relational.scala#L189-L192)

### CODE REVIEW

**Comment by danking** [comment]
@cseed is this how it should work? seems odd to put `null` as position. I also notice that `filterSamplesMask` does some custom stuff. Will this change slow things down unnecessarily if `drop_samples` is not used immediately after a read?

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala [hunk 1]
```diff
@@ -881,9 +881,7 @@ class VariantSampleMatrix[RPK, RK, T >: Null](val hc: HailContext, val metadata:
   def deleteVA(path: List[String]): (Type, Deleter) = vaSignature.delete(path)
 
   def dropSamples(): VariantSampleMatrix[RPK, RK, T] =
-    copy(sampleIds = IndexedSeq.empty[Annotation],
-      sampleAnnotations = IndexedSeq.empty[Annotation],
-      rdd = rdd.mapValues { case (va, gs) => (va, Iterable.empty[T]) })
+    copyAST(ast = FilterSamples(ast, Const(null, true, TBoolean)))
```
Comment (cseed) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: You want `false` here, right?

**Comment by cseed** [comment]
Yes.

 > seems odd to put null as position

I agree.  That's why we need your new IR with no positions!

 > Will this change slow things down unnecessarily if drop_samples is not used immediately after a read?

Yes, possibly a little, since FilterSamples needs to handle the case where not all samples are filtered.  Options: specialize the case in FilterSamples when the condition is false (or true -- no-op) or add a DropSamples node.  I prefer the former.

FilterSamples is somewhat better in the coming OrderedRDD2/RegionValueBuilder stuff.  The implementation has almost no overhead for when the condition is false.  I say leave this as is and we'll pick that up when my later PRs go in.

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala (hunk 1 above)
Comment (danking) [reply src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: I know every FP lang does this but I dislike the term `filter`. I always have to lookup if its a `select` or a `reject`.

**Pull request merged**